atom/renderer/electron_api_service_impl.cc

```diff
@@ -0,0 +1,151 @@
+// Copyright (c) 2019 Slack Technologies, Inc.
+// Use of this source code is governed by the MIT license that can be
+// found in the LICENSE file.
+
+#include "electron/atom/renderer/electron_api_service_impl.h"
+
+#include <utility>
+#include <vector>
+
+#include "atom/common/heap_snapshot.h"
+#include "atom/common/native_mate_converters/value_converter.h"
+#include "base/macros.h"
+#include "base/threading/thread_restrictions.h"
+#include "electron/atom/common/api/event_emitter_caller.h"
+#include "electron/atom/common/node_includes.h"
+#include "electron/atom/common/options_switches.h"
+#include "electron/atom/renderer/atom_render_frame_observer.h"
+#include "mojo/public/cpp/system/platform_handle.h"
+#include "native_mate/dictionary.h"
+#include "third_party/blink/public/web/blink.h"
+#include "third_party/blink/public/web/web_local_frame.h"
+
+namespace atom {
+
+namespace {
+
+const char kIpcKey[] = "ipcNative";
+
+void InvokeIpcCallback(v8::Handle<v8::Context> context,
+                       const std::string& callback_name,
+                       std::vector<v8::Handle<v8::Value>> args) {
+  TRACE_EVENT0("devtools.timeline", "FunctionCall");
```

Comment:
Seems like the common elements in `InvokeIpcCallback` and `InvokeHiddenCallback` can be extracted into some v8 util inside native mate

Reply:
I'd prefer to avoid modifications to native_mate, as I'm expecting it to eventually go away and be replaced by gin. I agree that these functions are largely similar though!